include/flang/ISO_Fortran_binding.h

```diff
@@ -45,11 +45,19 @@ typedef unsigned char CFI_rank_t;
 // and should be used for all array indexing and loop bound calculations.
 typedef ptrdiff_t CFI_index_t;
 
-#define CFI_DESC_T(rank) \
+#ifdef __cplusplus
+// C++ does not support zero-sized array (rank=0)
+// Also, CFI_cdesc_t already contains 1 dim in cpp
+#define CFI_CDESC_T(rank) \
+  std::aligned_storage_t<sizeof(CFI_cdesc_t) + (rank - 1) * sizeof(CFI_dim_t), \
```

Comment:
Why the `aligned_storage_t`?  Is the type `CFI_cdesc_t` not already aligned properly?

I am nervous about a potentially negative value being multiplied against an unsigned value.

Reply:
Maybe one could avoid the C++ restriction on zero-length arrays by changing the type of the array in that case to a zero-size struct, i.e. `std::conditional_t<rank != 0, CFI_dim_t, MyLocalEmptyStructType> dim[rank | !rank];`.

Reply:
Well, my issue was with the size not with the alignment. Because of the `dim[1]` inside the definition of `CFI_cdesc_t` (independent of the CFI_CDESC_T macro definition), we would waste 24bytes in the case of a C++ user defined scalar descriptor if the macro uses a structure containing `CFI_cdesc_t`. So I wanted to replacing it with a structure containing only a `char[]` to allocate what is necessary (similarly to the `new char[bytes]` used by the runtime in descriptor.cc for storage allocation).  Then I was unsure that we would get the alignement required by `CFI_cdesc_t`, an therefore used `aligne_storage_t`.
 
Thanks for the suggestion of `std::conditional_t`.  To push the reasoning on the space saving, this means we would have to use it also inside the `CFI_cdesc_t` type definition inside ISO_Fortran_binding.h. I think it would work though it is unclear to me what could be the consequences of having such conditional type when the standard says the `CFI_cdesc_t `structure dim member is a `CFI_dim_t[]`. I cannot imagine any user code that it would break.
Was your idea also to use such conditional type in `CFI_cdesc_t `definition or only in `CFI_CDESC_T `macro?

Reply:
If the rank is zero the code should never care about the type of the `dim[]` data member, although it might affect the behavior of the code in descriptor.h that locates the addendum.

The conditional type would not be used in the definition of `CFI_cdesc_t` because the predicate has to be testable at compilation time.

The other thing that you could do is just waste the 24 bytes in the case of a descriptor for a scalar in C++.